### ISSUE TITLE
nodetool cleanup causes Scylla to crash

### ISSUE THREAD

**Issue by r4fek** [scylladb/scylla#6263]
*Installation details*
Scylla version (or git commit hash): latest 3.2.3
Cluster size: 29
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian Stretch

I ran `nodetool cleanup` command on one of my nodes and this happened:

```
kwi 22 06:33:24 lati-dbs1 scylla[46722]:  [shard 15] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/user_device-2337510072f211e8809a00000000001e/mc-464477-big-Data.db:level=0, ]. 17761627 bytes to 13074731 (~73% of original) in 14181ms = 0.88MB/s. ~50304 total partitions merged to 36097.
kwi 22 06:33:24 lati-dbs1 scylla[46722]:  [shard 15] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-469215-big-Data.db:level=0, ]
kwi 22 06:33:25 lati-dbs1 scylla[46722]:  [shard 5] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/user_device-2337510072f211e8809a00000000001e/mc-472103-big-Data.db:level=0, ]. 18122171 bytes to 13096020 (~72% of original) in 10470ms = 1.19MB/s. ~54912 total partitions merged to 37324.
kwi 22 06:33:25 lati-dbs1 scylla[46722]:  [shard 5] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-476197-big-Data.db:level=0, ]
kwi 22 06:33:29 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-589742-big-Data.db:level=0, ]. 70923916 bytes to 70918340 (~99% of original) in 23628ms = 2.86MB/s. ~7040 total partitions merged to 6911.
kwi 22 06:33:29 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-671392-big-Data.db:level=0, ]
kwi 22 06:33:31 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-671438-big-Data.db:level=0, ]. 1020380 bytes to 532431 (~52% of original) in 1721ms = 0.30MB/s. ~256 total partitions merged to 164.
kwi 22 06:33:31 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaning [/var/lib/scylla/data/sync/synced_notification_by_id-0a13732094ad11e89e8000000000002b/mc-100026-big-Data.db:level=0, ]
kwi 22 06:33:32 lati-dbs1 scylla[46722]:  [shard 24] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-589146-big-Data.db:level=0, ]. 52983601 bytes to 52979034 (~99% of original) in 22373ms = 2.26MB/s. ~6016 total partitions merged to 5904.
kwi 22 06:33:32 lati-dbs1 scylla[46722]:  [shard 24] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-662930-big-Data.db:level=0, ]
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 41] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-580009-big-Data.db:level=0, ]. 93853109 bytes to 93842950 (~99% of original) in 30400ms = 2.94MB/s. ~8320 total partitions merged to 8245.
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 41] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-510227-big-Data.db:level=0, ]
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/synced_notification_by_id-0a13732094ad11e89e8000000000002b/mc-100072-big-Data.db:level=0, ]. 761593 bytes to 658240 (~86% of original) in 2364ms = 0.27MB/s. ~16384 total partitions merged to 13902.
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaning [/var/lib/scylla/data/sync/synced_notification-064f3f3094ad11e89e8000000000002b/mc-100256-big-Data.db:level=0, ]
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 12] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-564478-big-Data.db:level=0, ]. 52726063 bytes to 52719281 (~99% of original) in 24926ms = 2.02MB/s. ~6016 total partitions merged to 5896.
kwi 22 06:33:33 lati-dbs1 scylla[46722]:  [shard 12] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-640148-big-Data.db:level=0, ]
kwi 22 06:33:35 lati-dbs1 scylla[46722]:  [shard 36] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/typed_url_by_id-0c498d0094ad11e893cf000000000017/mc-577980-big-Data.db:level=0, ]. 93365678 bytes to 93357595 (~99% of original) in 31721ms = 2.81MB/s. ~8064 total partitions merged to 7973.
kwi 22 06:33:35 lati-dbs1 scylla[46722]:  [shard 36] compaction - Cleaning [/var/lib/scylla/data/sync/typed_url-0b1cccd094ad11e88e02000000000025/mc-654432-big-Data.db:level=0, ]
kwi 22 06:33:35 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaned 1 sstables to [/var/lib/scylla/data/sync/synced_notification-064f3f3094ad11e89e8000000000002b/mc-100302-big-Data.db:level=0, ]. 761884 bytes to 658492 (~86% of original) in 1978ms = 0.32MB/s. ~16384 total partitions merged to 13901.
kwi 22 06:33:35 lati-dbs1 scylla[46722]:  [shard 22] compaction - Cleaning [/var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/mc-675946-big-Data.db:level=2, ]
kwi 22 06:33:37 lati-dbs1 scylla[46722]: scylla: sstables/mp_row_consumer.hh:433: row_consumer::proceed sstables::mp_row_consumer_k_l::flush(): Assertion `!_ready' failed.
kwi 22 06:38:40 lati-dbs1 systemd[1]: scylla-server.service: Main process exited, code=killed, status=6/ABRT
kwi 22 06:38:41 lati-dbs1 systemd[1]: scylla-server.service: Unit entered failed state.
kwi 22 06:38:41 lati-dbs1 systemd[1]: scylla-server.service: Failed with result 'signal'.
kwi 22 06:38:41 lati-dbs1 systemd[1]: scylla-server.service: Service hold-off time over, scheduling restart.
kwi 22 06:38:41 lati-dbs1 systemd[1]: Stopped Scylla Server.
kwi 22 06:38:41 lati-dbs1 systemd[1]: Starting Scylla Server...
```

**Comment by dorlaor** [comment]
@haaawk it's your lines ;)

**Comment by haaawk** [comment]
Looking...

**Comment by haaawk** [comment]
@r4fek This is a codepath for sstables KA/LA. I thought you're on MC, no?

**Comment by r4fek** [comment]
Yes, I'm using MC format.

**Comment by haaawk** [comment]
This is very strange then. The names of sstables in the log indicate they are MC sstables but the codepath is for KA/LA. Maybe nodetool cleanup handling has a bug. Let me see.

**Comment by haaawk** [comment]
@r4fek two more questions:
1. What's the nodetool parameters you've used? keyspace name or table name?
2. Do you have any KA/LA sstable files in keyspaces you've been cleaning up?

**Comment by r4fek** [comment]
1. just regular `nodetool cleanup`
2. No KA, but some LA leftovers:
```
root@lati-dbs1:/var/lib/scylla/data# find . -iname la-*
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Filter.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Scylla.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-TOC.txt
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Data.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Index.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Statistics.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-CompressionInfo.db
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Digest.sha1
./sync/entity2-1ee534f072f211e886a8000000000014/la-4935-big-Summary.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-TOC.txt
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Data.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Index.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Digest.sha1
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-CRC.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Summary.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Filter.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Statistics.db
./sync/bookmark-eab1b28094ac11e89e8000000000002b/la-220010-big-Scylla.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-TOC.txt
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Data.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Index.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Digest.sha1
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-CRC.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Summary.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Filter.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Statistics.db
./sync/session-0384385094ad11e88e02000000000025/la-186181-big-Scylla.db
```

**Comment by haaawk** [comment]
Any LA sstables in sync/session_by_id?

**Comment by r4fek** [comment]
No :(

**Comment by haaawk** [comment]
Would you be able to retry the nodetool cleanup with parameters pointing to a single table in sync keyspace so that we could narrow down the problem to a particular table?

sync/entity2, sync/bookmark and sync/session are the main suspects.

**Comment by psarna** [comment]
@haaawk we already had an issue with ka/la sstables appearing on startup, e.g. here: https://github.com/scylladb/scylla/issues/6071#issuecomment-603178343 . Maybe it's the source of old sstables appearing here as well.

**Comment by haaawk** [comment]
I don't think this is the same @psarna. I suspect that the problem is in cleanup operation when it mixes LA and MC but will have to find the prove.

**Comment by haaawk** [comment]
I wasn't able to find any flaw in cleanup code so I guess maybe it started cleaning one of the LA sstables and the bug is inside reading LA sstables. It's either that or some mix-up that directed MC sstable to LA codepath. There's one way which could rule out the first option. @r4fek could you please check in the logs whether there was any `compaction - Cleaning [/var/lib/...` entry for an LA sstable? If not then probably the problem is with MC sstables. Otherwise we can focus on KA/LA codepath and your old KA/LA sstables.

**Comment by r4fek** [comment]
@haaawk no, I don't have such entries in my log (cleaning la-).

**Comment by haaawk** [comment]
Thanks, @r4fek. I was hoping you do ;). Is the nodetool cleanup failure reproducible?

**Comment by r4fek** [comment]
Yes, I just crashed the node again ;)

**Comment by haaawk** [comment]
@r4fek Could you please try running `nodetool cleanup sync <table name>` for each table you have in the `sync` keyspace? I don't know how many you have there. If the number is big we could try bisecting the list of tables into two equal parts and run `nodetool cleanup sync <list of space-separated table names>` on one of them and if it breaks split it again. That would hopefully narrow down the problem to a single table.

**Comment by r4fek** [comment]
On it!

**Comment by haaawk** [comment]
BTW if it's ok for you to give me ssh access, I can do it for you.

**Comment by r4fek** [comment]
Not possible, sorry :) I'll get back to you with problematic table name shortly.

**Comment by haaawk** [comment]
> Not possible, sorry :) I'll get back to you with problematic table name shortly.

I understand. No problem.

**Comment by r4fek** [comment]
@haaawk The faulty table is `typed_url_by_id`.

**Comment by r4fek** [comment]
I have a few ka's here:
```
  494,5 MiB [#####     ]  sync-typed_url_by_id-ka-414166-Data.db
  485,7 MiB [#####     ]  sync-typed_url_by_id-ka-456512-Data.db
  469,4 MiB [####      ]  sync-typed_url_by_id-ka-479766-Data.db
  465,2 MiB [####      ]  sync-typed_url_by_id-ka-443495-Data.db
```

**Comment by r4fek** [comment]
I ran `nodetool compact sync password_by_id` (it migrated all KA leftovers to MC) and now `nodetool cleanup sync password_by_id` doesn't cause any crash.

**Comment by r4fek** [comment]
On other node some backtrace appeared before the restart:
```
/usr/lib/debug/.build-id/18/6045664da749e2c77b109834b4e13e8dd60f28.debug: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /opt/scylladb/libreloc/ld.so, for GNU/Linux 3.2.0, BuildID[sha1]=186045664da749e2c77b109834b4e13e8dd60f28, not stripped

kwi 28 07:27:12 lati-dbs2 scylla[37127]: Backtrace:
kwi 28 07:27:12 lati-dbs2 scylla[37127]:
[Backtrace #0]
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/c++/9/ext/new_allocator.h:105
seastar::backtrace_buffer::append_backtrace() at /usr/include/c++/9/ext/new_allocator.h:105
 (inlined by) print_with_backtrace at /jenkins/workspace/scylla-3.2/build/scylla/seastar/src/core/reactor.cc:731
seastar::print_with_backtrace(char const*) at /usr/include/c++/9/ext/new_allocator.h:105
sigabrt_action at /usr/include/c++/9/ext/new_allocator.h:105
 (inlined by) operator() at /jenkins/workspace/scylla-3.2/build/scylla/seastar/src/core/reactor.cc:3390
 (inlined by) _FUN at /jenkins/workspace/scylla-3.2/build/scylla/seastar/src/core/reactor.cc:3386
?? ??:0
/opt/scylladb/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=c48678dd67525a968f80fbf579203e7b4a9cddb5, for GNU/Linux 3.2.0, stripped, too many notes (256)

gsignal at ??:?
abort at ??:?
?? ??:0
__assert_fail at ??:?
sstables::mp_row_consumer_k_l::flush() at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/future.hh:300
sstables::mp_row_consumer_k_l::consume_row_end() at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/future.hh:300
 (inlined by) sstables::data_consume_rows_context::do_process_state(seastar::temporary_buffer<char>&) at /jenkins/workspace/scylla-3.2/build/scylla/sstables/row.hh:369
sstables::data_consume_rows_context::process_state(seastar::temporary_buffer<char>&) at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/future.hh:300
 (inlined by) data_consumer::continuous_data_consumer<sstables::data_consume_rows_context>::process(seastar::temporary_buffer<char>&) at /jenkins/workspace/scylla-3.2/build/scylla/sstables/consumer.hh:404
 (inlined by) data_consumer::continuous_data_consumer<sstables::data_consume_rows_context>::operator()(seastar::temporary_buffer<char>) at /jenkins/workspace/scylla-3.2/build/scylla/sstables/consumer.hh:437
 (inlined by) seastar::future<seastar::consumption_result<char> > std::__invoke_impl<seastar::future<seastar::consumption_result<char> >, sstables::data_consume_rows_context&, seastar::temporary_buffer<char> >(std::__invoke_other, sstables::data_consume_rows_context&, seastar::temporary_buffer<char>&&) at /usr/include/c++/9/bits/invoke.h:60
 (inlined by) std::__invoke_result<sstables::data_consume_rows_context&, seastar::temporary_buffer<char> >::type std::__invoke<sstables::data_consume_rows_context&, seastar::temporary_buffer<char> >(sstables::data_consume_rows_context&, seastar::temporary_buffer<char>&&) at /usr/include/c++/9/bits/invoke.h:96
 (inlined by) std::result_of<sstables::data_consume_rows_context& (seastar::temporary_buffer<char>&&)>::type std::reference_wrapper<sstables::data_consume_rows_context>::operator()<seastar::temporary_buffer<char> >(seastar::temporary_buffer<char>&&) const at /usr/include/c++/9/bits/refwrap.h:339
 (inlined by) seastar::future<> seastar::input_stream<char>::consume<std::reference_wrapper<sstables::data_consume_rows_context> >(std::reference_wrapper<sstables::data_consume_rows_context>&&)::{lambda()#1}::operator()() at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/iostream-impl.hh:227
 (inlined by) seastar::future<seastar::bool_class<seastar::stop_iteration_tag> > seastar::futurize<seastar::future<seastar::bool_class<seastar::stop_iteration_tag> > >::apply<seastar::future<> seastar::input_stream<char>::consume<std::reference_wrapper<sstables::data_consume_rows_context> >(std::reference_wrapper<sstables::data_consume_rows_context>&&)::{lambda()#1}&>(std::reference_wrapper<sstables::data_consume_rows_context>&&) at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/future.hh:1547
 (inlined by) seastar::future<> seastar::repeat<seastar::future<> seastar::input_stream<char>::consume<std::reference_wrapper<sstables::data_consume_rows_context> >(std::reference_wrapper<sstables::data_consume_rows_context>&&)::{lambda()#1}>(seastar::future<> seastar::input_stream<char>::consume<std::reference_wrapper<sstables::data_consume_rows_context> >(std::reference_wrapper<sstables::data_consume_rows_context>&&)::{lambda()#1}) at /jenkins/workspace/scylla-3.2/build/scylla/seastar/include/seastar/core/future-util.hh:337
```

**Comment by slivne** [comment]
@haaawk ping

**Comment by slivne** [comment]
nodetool compact indeed moved all to mc but may have also removed tombstones as well.

**Comment by tgrabiec** [comment]
Caused by  scylladb/scylla#7553

**Comment by avikivity** [comment]
@tgrabiec is it a duplicate? If so we can close.

**Comment by tgrabiec** [comment]
Fixed in fb9b5ca
